### ISSUE TITLE
Recursive TS-transpiled async function crashes Duktape

### ISSUE THREAD

**Issue by fatcerberus** [svaarala/duktape#1384]
I'm assuming this is a stack overflow.

The following TypeScript:
```ts
async function explode()
{
	await explode();
}

explode();
```

Produces this (extremely verbose :) JS:
```js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
function explode() {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, explode()];
                case 1:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    });
}
explode();
```

Which when run (providing a Promise implementation is available--I use babel-polyfill/core-js), causes a segfault.

**Comment by svaarala** [comment]
Just checking: this is a segfault and not an uncaught error? Can this be repeated by running 'duk' with the Promise implementation and the snippet above?

**Comment by svaarala** [comment]
Can't seem to trigger the issue with Bluebird promise implementation on 2.0.x or master.

**Comment by svaarala** [comment]
Running with Bluebird promises and `dukd` (assertions and debug prints) seems to trigger some issue, possibly a C stack overflow?

**Comment by svaarala** [comment]
Interestingly `duk` doesn't trigger the issue (for me), `dukd` does, but the issue triggered is not an assert failure but a segfault. Valgrind suggests that a stack overflow might be the cause; increasing valgrind stack size from 8MB to 16MB fixes the (valgrind) issue for `dukd` too.

Could this be a C stack overflow in your environment? Could you increase the C stack size to see if that affects the issue?

**Comment by svaarala** [comment]
(Ahem, I see that you mentioned stack overflow as a potential cause right in the description... Sorry :-)

**Comment by fatcerberus** [comment]
I suspected a stack overflow, but wasn't sure because this is just JS-to-JS calls, which shouldn't grow the C stack from my understanding of the bytecode executor.  Unless there's a `.call()` or `.apply()` hidden in there that I'm not seeing; that could explain it if so.

**Comment by svaarala** [comment]
There's something in the ever growing callstack which does require a native recursive call:

```
D0 duk_js_call.c:2406 (duk_handle_ecma_call_setup): final target is a lightfunc/nativefunc, cannot do ecma-to-ecma call
D0 duk_js_call.c:2412 (duk_handle_ecma_call_setup): ecma-to-ecma OK
D0 duk_js_call.c:2412 (duk_handle_ecma_call_setup): ecma-to-ecma OK
D0 duk_js_call.c:2412 (duk_handle_ecma_call_setup): ecma-to-ecma OK
D0 duk_js_call.c:2412 (duk_handle_ecma_call_setup): ecma-to-ecma OK
D0 duk_js_call.c:2412 (duk_handle_ecma_call_setup): ecma-to-ecma OK
D0 duk_js_call.c:2412 (duk_handle_ecma_call_setup): ecma-to-ecma OK
D0 duk_js_call.c:2412 (duk_handle_ecma_call_setup): ecma-to-ecma OK
D0 duk_js_call.c:2412 (duk_handle_ecma_call_setup): ecma-to-ecma OK
D0 duk_js_call.c:2406 (duk_handle_ecma_call_setup): final target is a lightfunc/nativefunc, cannot do ecma-to-ecma call
D0 duk_js_call.c:2412 (duk_handle_ecma_call_setup): ecma-to-ecma OK
D0 duk_js_call.c:2412 (duk_handle_ecma_call_setup): ecma-to-ecma OK
D0 duk_js_call.c:2412 (duk_handle_ecma_call_setup): ecma-to-ecma OK
D0 duk_js_call.c:2412 (duk_handle_ecma_call_setup): ecma-to-ecma OK
D0 duk_js_call.c:2412 (duk_handle_ecma_call_setup): ecma-to-ecma OK
D0 duk_js_call.c:2412 (duk_handle_ecma_call_setup): ecma-to-ecma OK
D0 duk_js_call.c:2406 (duk_handle_ecma_call_setup): final target is a lightfunc/nativefunc, cannot do ecma-to-ecma call
D0 duk_js_call.c:2412 (duk_handle_ecma_call_setup): ecma-to-ecma OK
D0 duk_js_call.c:2412 (duk_handle_ecma_call_setup): ecma-to-ecma OK
D0 duk_js_call.c:2412 (duk_handle_ecma_call_setup): ecma-to-ecma OK
D0 duk_js_call.c:2412 (duk_handle_ecma_call_setup): ecma-to-ecma OK
D0 duk_js_call.c:2412 (duk_handle_ecma_call_setup): ecma-to-ecma OK
D0 duk_js_call.c:2412 (duk_handle_ecma_call_setup): ecma-to-ecma OK
D0 duk_js_call.c:2412 (duk_handle_ecma_call_setup): ecma-to-ecma OK
D0 duk_js_call.c:2412 (duk_handle_ecma_call_setup): ecma-to-ecma OK
D0 duk_js_call.c:2406 (duk_handle_ecma_call_setup): final target is a lightfunc/nativefunc, cannot do ecma-to-ecma call
D0 duk_js_call.c:2412 (duk_handle_ecma_call_setup): ecma-to-ecma OK
D0 duk_js_call.c:2412 (duk_handle_ecma_call_setup): ecma-to-ecma OK
D0 duk_js_call.c:2412 (duk_handle_ecma_call_setup): ecma-to-ecma OK
D0 duk_js_call.c:2412 (duk_handle_ecma_call_setup): ecma-to-ecma OK
D0 duk_js_call.c:2412 (duk_handle_ecma_call_setup): ecma-to-ecma OK
D0 duk_js_call.c:2412 (duk_handle_ecma_call_setup): ecma-to-ecma OK
D0 duk_js_call.c:2406 (duk_handle_ecma_call_setup): final target is a lightfunc/nativefunc, cannot do ecma-to-ecma call
```

The native call seems to be `.call()` inside step().

**Comment by svaarala** [comment]
If this is just an "ordinary" stack overflow, there are two issues:

- Default stack limits are not low enough for this particular case on Windows.
- Ideally both .call() and .apply() would support an Ecma-to-Ecma call without native recursion.

**Comment by svaarala** [comment]
Added #1388 to track direct handling of .call() and .apply(), a known future work issue in call handling.

**Comment by fatcerberus** [comment]
Ah, I see there are two `.call()`s in there now.  That's probably the cause, but I'll verify in the debugger when I get in later to be sure it's indeed just a regular stack overflow and not something else.

Off topic: The fact that the TS transpiler basically generated an entire stack machine for my little generator function is quite interesting.  Who knew it was possible to emulate async/await in ES5? :)

**Comment by fatcerberus** [comment]
By the way, I believe Windows only allocates a static 8MB of stack for each process by default--more than that must be requested at compile time if needed and the OS will not grow the stack automatically.

**Comment by fatcerberus** [comment]
Actually it's worse than that--all the sources I can find say MSVC uses a default of 1MB for the stack.  Ouch.

**Comment by svaarala** [comment]
Yeah. It's quite difficult to get the default limits (for call stacks and anything else affecting native recursion) right with respect to platform stack limits. There are two moving targets: one is the size of stack frames which changes over time due to Duktape and compiler changes, and another is the platform stack initialization behavior.

I don't really see a workable solution using fixed limits, which is why I think #231 with maybe a default stack check included for mainstream platforms.

**Comment by svaarala** [comment]
> Who knew it was possible to emulate async/await in ES5? :)

Being Turing complete, ES5 can ultimately emulate anything, it's just a question of efficiency :) But yeah, it's interesting that the transpiled code looks like that.

**Comment by fatcerberus** [comment]
Ran the program several times under MSVC; the first time I got an access violation during a `malloc()` call, which made me think heap corruption... however, all subsequent runs produced a stack overflow, as expected.

**Comment by fatcerberus** [comment]
Increasing the stack size to 16MB seems to fix the segfault in all cases *except* when the Duktape debugger is connected.

Oddly enough `explode()` doesn't cause an infinite loop and the program continues after the call.  I assume that's because as an async function it runs "in the background" via promise resolution.

**Comment by fatcerberus** [comment]
Okay, that's kind of weird.  I added a `console.log()` to log how many times the recursive function is reentered, and it apparently gives up completely after 332 iterations (this is with the 16MB C stack).  With the debugger connected, it still makes it to 332 iterations, but then it segfaults.

**Comment by fatcerberus** [comment]
With the 16MB stack, I found this in my console output:
```
ERROR: Unhandled promise rejection RangeError: C call stack depth limit
```

Which is interesting, because async functions operate via continuations through promise resolution--there should be no C recursion at all.

**Comment by svaarala** [comment]
But like discussed above, there's a recursion which involves a .call(), which then increases the C recursion depth? Or maybe I misunderstood something?

**Comment by fatcerberus** [comment]
Yes, I'm just trying to figure out why the .call() *is* infinitely recursive.  The `await` should cause the generator to yield, and not continue until the promise is settled.

...although now that I look at it, each promise is contingent on another one deeper in the recursion, ad infinitum.  Promise construction itself is synchronous.  So forget I said anything. :)
